facilitator/src/task/pubsub.rs

```diff
@@ -0,0 +1,187 @@
+/// parsed here, only the ones used by this application.
+/// https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+struct PubSubMessage {
+    data: String,
+    message_id: String,
+    publish_time: String,
+}
+
+/// A task queue backed by Google Cloud PubSub
+#[derive(Debug)]
+pub struct PubSubTaskQueue<T: Task> {
+    pubsub_api_endpoint: String,
+    gcp_project_id: String,
+    subscription_id: String,
+    oauth_token_provider: OauthTokenProvider,
+    phantom_task: PhantomData<*const T>,
+}
+
+impl<T: Task> PubSubTaskQueue<T> {
+    pub fn new(
+        pubsub_api_endpoint: Option<&str>,
+        gcp_project_id: &str,
+        subscription_id: &str,
+        identity: Identity,
+    ) -> Result<PubSubTaskQueue<T>> {
+        Ok(PubSubTaskQueue {
+            pubsub_api_endpoint: pubsub_api_endpoint
+                .unwrap_or(PUBSUB_API_BASE_URL)
+                .to_owned(),
+            gcp_project_id: gcp_project_id.to_string(),
+            subscription_id: subscription_id.to_string(),
+            oauth_token_provider: OauthTokenProvider::new(
+                // This token is used to access PubSub API
+                // https://developers.google.com/identity/protocols/oauth2/scopes
+                "https://www.googleapis.com/auth/pubsub",
+                identity.map(|x| x.to_string()),
+                None, // GCP key file; never used
+            )?,
+            phantom_task: PhantomData,
+        })
+    }
+}
+
+impl<T: Task> TaskQueue<T> for PubSubTaskQueue<T> {
+    fn dequeue(&mut self) -> Result<Option<TaskHandle<T>>> {
+        info!(
+            "pull task from {}/{} as {:?}",
+            self.gcp_project_id, self.subscription_id, self.oauth_token_provider
+        );
+
+        // API reference: https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull
+        let url = format!(
+            "{}/v1/projects/{}/subscriptions/{}:pull",
+            self.pubsub_api_endpoint, self.gcp_project_id, self.subscription_id
+        );
+
+        let http_response = ureq::post(&url)
+            .set(
+                "Authorization",
+                &format!("Bearer {}", self.oauth_token_provider.ensure_oauth_token()?),
+            )
+            // By default, ureq will wait forever to connect or read.
+            // Empirically, if there are no messages available in the
+            // subscription, the PubSub API will wait about 90 seconds to send
+            // an HTTP 200 with an empty JSON body. We set a higher read timeout
+            // than usual to allow for this.
+            .timeout_connect(10_000) // ten seconds
+            .timeout_read(180_000) // three minutes
+            .set("Content-Type", "application/json")
+            .send_json(ureq::json!({
+                // Dequeue one task at a time
+                "maxMessages": 1
+            }));
```

Comment:
This repeated block (shared with function below) should be factored out into `self.send_json(read_timeout, json)`

Reply:
I refactored this into the `send_json_request` function in the `http` module. There's more boilerplate around `ureq` usage that could be cleaned up, but I think step #1 there is to move to `ureq` > 2.0.0, and I don't want to do that in this PR, so I filed #297